drivers/dai/intel/alh/alh.c

```diff
@@ -0,0 +1,223 @@
+{
+	struct dai_config *params = (struct dai_config *)dev->config;
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+
+	params->rate = alh->params.rate;
+	params->channels = alh->params.channels;
+	params->word_size = ALH_WORD_SIZE_DEFAULT;
+
+	return params;
+}
+
+static int dai_alh_config_set(const struct device *dev, const struct dai_config *cfg,
+				  const void *bespoke_cfg)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+
+	LOG_INF("%s", __func__);
+
+	alh_claim_ownership();
+
+	if (cfg->type == DAI_INTEL_ALH) {
+		return dai_alh_set_config_tplg(dp, cfg, bespoke_cfg);
+	} else {
+		return dai_alh_set_config_blob(dp, bespoke_cfg);
+	}
+}
+
+static const struct dai_properties *dai_alh_get_properties(const struct device *dev,
+							   enum dai_dir dir, int stream_id)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+	struct dai_properties *prop = &alh->props;
+	uint32_t offset = dir == DAI_DIR_PLAYBACK ?
+		ALH_TXDA_OFFSET : ALH_RXDA_OFFSET;
+
+	prop->fifo_address = dai_base(dp) + offset + ALH_STREAM_OFFSET * stream_id;
+	prop->dma_hs_id = alh_handshake_map[stream_id];
+	prop->stream_id = alh->params.stream_id;
+
+	LOG_INF("%s dai_index %u", __func__, dp->index);
+	LOG_INF("%s fifo %u", __func__, prop->fifo_address);
+	LOG_INF("%s handshake %u", __func__, prop->dma_hs_id);
+
+	return prop;
+}
+
+static int dai_alh_probe(struct dai_intel_alh *dp)
+{
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+
+	LOG_INF("%s", __func__);
+
+	memset(alh, 0, sizeof(*alh));
+
+	return 0;
+}
+
+static int dai_alh_remove(struct dai_intel_alh *dp)
+{
+	LOG_INF("%s", __func__);
+
+	alh_release_ownership();
+
+	return 0;
+}
+
+static int dai_alh_probe_wrapper(const struct device *dev)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
+
+	key = k_spin_lock(&dp->lock);
+
+	if (dp->sref == 0) {
+		ret = dai_alh_probe(dp);
+	}
+
+	if (!ret) {
+		dp->sref++;
+	}
+
+	k_spin_unlock(&dp->lock, key);
+
+	return ret;
+}
+
+static int dai_alh_remove_wrapper(const struct device *dev)
+{
+	struct dai_intel_alh *dp = (struct dai_intel_alh *)dev->data;
+	k_spinlock_key_t key;
+	int ret = 0;
+
+	key = k_spin_lock(&dp->lock);
+
+	if (--dp->sref == 0) {
+		ret = dai_alh_remove(dp);
+	}
+
+	k_spin_unlock(&dp->lock, key);
+
+	return ret;
+}
+
+static int alh_init(const struct device *dev)
+{
+	return 0;
+}
+
+static struct dai_driver_api dai_intel_alh_api_funcs = {
+	.probe			= dai_alh_probe_wrapper,
+	.remove			= dai_alh_remove_wrapper,
+	.config_set		= dai_alh_config_set,
+	.config_get		= dai_alh_config_get,
+	.trigger		= dai_alh_trigger,
+	.get_properties		= dai_alh_get_properties,
+};
+
+#define DAI_INTEL_ALH_DEVICE_INIT(n)						\
+	static struct dai_config dai_intel_alh_config_##n;			\
```

Comment:
Per comment earlier: traditionally the "config" struct for a zephyr device is declared const and contains only immutable data.  Dynamic state goes in the "data" struct.

Reply:
This way it was upstream with dai ssp, I think this is another thing to change in dai refactor because for now dai_config is defined in dai and cannot be changed only for alh